core/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java

```diff
@@ -147,8 +147,17 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
         RoutingNodes routingNodes = allocation.routingNodes();
 
         // First, handle primaries, they must find a place to be allocated on here
-        MetaData metaData = routingNodes.metaData();
-        Iterator<ShardRouting> unassignedIterator = routingNodes.unassigned().iterator();
+        final MetaData metaData = routingNodes.metaData();
+        RoutingNodes.UnassignedShards unassigned = routingNodes.unassigned();
+        unassigned.sort(new PriorityComparator() {
```

Comment:
We could avoid the anonymous class by just passing `metaData` to `PriorityComparator` and doing the work there. It seems like we used the anonymous class to avoid a constructor, but I recognize this is largely style.